llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp

```diff
@@ -1688,174 +1696,426 @@ bool PreRARematStage::allUsesAvailableAt(const MachineInstr *InstToRemat,
+       << Excess.ArchVGPRsToAlignment << " registers)\n";
+    return OS;
+  }
+#endif
 
-    MachineOperand *Op = DAG.MRI.getOneDef(Reg);
-    MachineInstr *Def = Op->getParent();
-    if (Op->getSubReg() != 0 || !isTriviallyReMaterializable(*Def))
-      continue;
+private:
+  static inline bool saveRegs(unsigned &LeftToSave, unsigned &NumRegs) {
+    unsigned NumSaved = std::min(LeftToSave, NumRegs);
+    NumRegs -= NumSaved;
+    LeftToSave -= NumSaved;
+    return NumSaved;
+  }
+};
+} // namespace
+
+ExcessRP::ExcessRP(const GCNSubtarget &ST, const GCNRegPressure &RP,
+                   unsigned MaxVGPRs) {
+  unsigned NumArchVGPRs = RP.getArchVGPRNum();
+  unsigned NumAGPRs = RP.getAGPRNum();
+  HasAGPRs = NumAGPRs;
+
+  if (!ST.hasGFX90AInsts()) {
+    // Non-unified RF. Account for excess pressure for ArchVGPRs and AGPRs
+    // independently.
+    if (NumArchVGPRs > MaxVGPRs)
+      ArchVGPRs = NumArchVGPRs - MaxVGPRs;
+    if (NumAGPRs > MaxVGPRs)
+      AGPRs = NumAGPRs - MaxVGPRs;
+    return;
+  }
 
-    MachineInstr *UseI = &*DAG.MRI.use_instr_nodbg_begin(Reg);
-    if (Def->getParent() == UseI->getParent())
-      continue;
+  // Independently of whether overall VGPR pressure is under the limit, we still
+  // have to check whether ArchVGPR pressure or AGPR pressure alone exceeds the
+  // number of addressable registers in each category.
+  const unsigned MaxArchVGPRs = ST.getAddressableNumArchVGPRs();
+  if (NumArchVGPRs > MaxArchVGPRs) {
+    ArchVGPRs = NumArchVGPRs - MaxArchVGPRs;
+    NumArchVGPRs = MaxArchVGPRs;
+  }
+  if (NumAGPRs > MaxArchVGPRs) {
+    AGPRs = NumAGPRs - MaxArchVGPRs;
+    NumAGPRs = MaxArchVGPRs;
+  }
 
-    bool HasRematDependency = false;
-    // Check if this instruction uses any registers that are planned to be
-    // rematerialized
-    for (auto &RematEntry : RematerializableInsts) {
-      if (find_if(RematEntry.second,
-                  [&Def](std::pair<MachineInstr *, MachineInstr *> &Remat) {
-                    for (MachineOperand &MO : Def->operands()) {
-                      if (!MO.isReg())
-                        continue;
-                      if (MO.getReg() == Remat.first->getOperand(0).getReg())
-                        return true;
-                    }
-                    return false;
-                  }) != RematEntry.second.end()) {
-        HasRematDependency = true;
-        break;
-      }
+  // Check overall VGPR usage against the limit; any excess above addressable
+  // register limits has already been accounted for.
+  unsigned NumVGPRs = GCNRegPressure::getUnifiedVGPRNum(NumArchVGPRs, NumAGPRs);
+  if (NumVGPRs > MaxVGPRs) {
+    VGPRs = NumVGPRs - MaxVGPRs;
+    ArchVGPRsToAlignment = NumArchVGPRs - alignDown(NumArchVGPRs, 4);
+    if (!ArchVGPRsToAlignment)
+      ArchVGPRsToAlignment = 4;
+  }
+}
+
+bool ExcessRP::saveArchVGPRs(unsigned NumRegs, bool UseArchVGPRForAGPRSpill) {
+  bool Progress = saveRegs(ArchVGPRs, NumRegs);
```

Comment:
For many cases `NumRegs` will be 0 after the call to `saveRegs`. The only way I can make sense of the below handling is if `NumRegs` maintained it original value after the call to `saveRegs` -- for example, if our remat reduced VGPR pressure by 4, we should reduce both the `ExcessRP.ArchVGPRs` and `Excess.VGPRs` by this amount. Maybe `saveRegs` shouldn't take `NumRegs` by reference?

Reply:
I think this is ok because `ExcessRP.ArchVGPRs` only tracks VGPR savings that must specifically be ArchVGPRs (because of the addressable limit). Once this has been reduced to 0 (e.g., on gfx942 once we have ArchVGPR usage <= 256) we start to use ArchVGPR savings as "generic VGPR savings".

Taking my above example again. On the first remat of an ArchVGPR-defining instruction (`NumRegs == 1`), `ArchVGPRs` will be reduced to 0 by this call and so will `NumRegs`, making the rest of the function void. On subsequent calls to `saveArchVGPRs`, this will change nothing to its arguments (so `NumRegs == 1` after the call) and return false, so we will hit `Progress |= saveRegs(VGPRs, NumRegs)` and start reducing the number of "generic" excess `VGPRs`.

Reply:
Oh I see -- you are not using the the ArchVGPR usage in calculating the VGPR usage, but are using instead the ArchVGPR target (i.e. MaxArchVGPRs).